### PR TITLE
feat(resolver): add new APIs for Circle features

### PR DESCRIPTION
**Pull request by devformatters**

### CODE REVIEW

File: src/connectors/atomService.ts [hunk 1]
```diff
@@ -29,27 +49,114 @@ interface DeleteManyInput {
...
+  draftIdLoader: DataLoader<string, Item>
+  userIdLoader: DataLoader<string, Item>
+
   constructor() {
     super()
     this.es = es
+    this.aws = aws
     this.knex = knex
+
+    this.circleIdLoader = this.initLoader({ table: 'circle', mode: 'id' })
```
Comment (guoliu) [inline src/connectors/atomService.ts:75]: Do you think it's better to implement the dataloaders here, or should we allow resolvers to pass in table name and create their own dataloader (for example with `initLoader`)? cc: @robertu7

**Review by guoliu — APPROVED** [review]
Left a comment, else looks good! 🍀

File: src/queries/article/circle.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+const resolver: ArticleToCircleResolver = async (
+  { id },
+  _,
+  { dataSources: { atomService } }
+) => {
+  if (!id) {
+    return []
+  }
+
+  const circleIds = await atomService.findMany({
```
Comment (robertu7) [inline src/queries/article/circle.ts:12]: find first?

File: src/queries/circle/isMember.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+const resolver: CircleToIsMemberResolver = async (
+  { id },
+  _,
+  { viewer, dataSources: { atomService } }
+) => {
+  if (!viewer.id) {
+    return false
+  }
+
+  const record = await atomService.findFirst({
```
Comment (robertu7) [inline src/queries/circle/isMember.ts:13]: same as the comment above, subscription item as the source of truth?

**Review by robertu7 — CHANGES_REQUESTED** [review]

File: src/connectors/atomService.ts (hunk 1 above)
Comment (devformatters) [reply src/connectors/atomService.ts:75]: 😉 Have no strong preference to it, but I guess we'll have redundant loaders easily. For example, we need `article`, `comment`, `user` loaders when resolving `User`, and also need `comment`, `draft` and `user` loaders when resolving `Article`.

File: src/types/circle.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    "Whther discussion is enabled or not."
+    enableDiscussion: Boolean!
+  }
+
+  type Member {
+    "User who join to a Circle."
+    user: User! @logCache(type: "${NODE.user}")
+
+    "Price chosen by user when joining a Circle."
+    price: Price!
```
[the remark this answers is not in the document]
Comment (devformatters) [reply src/types/circle.ts:95]: You're right, but we won't implement invitation mechanism in MVP, right?

File: src/types/circle.ts [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+  extend type Query {
+    circle(input: CircleInput!): Circle @privateCache @logCache(type: "${NODE.circle}")
+  }
+
+  extend type Mutation {
+    "Put a Circle."
+    putCircle(input: PutCircleInput!): Circle! @auth(mode: "${MODE.oauth}", group: "${GROUP.level3}") @purgeCache(type: "${NODE.circle}")
+
+    "Add or remove a Circle member."
+    toggleCircleMember(input: ToggleCircleMemberInput!): Circle! @auth(mode: "${MODE.oauth}", group: "${GROUP.level3}") @purgeCache(type: "${NODE.circle}")
```
[the remark this answers is not in the document]
Comment (devformatters) [reply src/types/circle.ts:18]: Yes, i just put a container here, but you're right.

File: src/mutations/circle/putCircle.ts [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+      const [hasCircle, sameCircle] = await Promise.all([
+        atomService.count({ table: 'circle', where: { owner: viewer.id } }),
+        atomService.count({
+          table: 'circle',
+          where: { name: trimedName },
+        }),
+      ])
+
+      if (hasCircle > 0) {
+        throw new DuplicateCircleError('alredy own a circle')
```
[the remark this answers is not in the document]
Comment (devformatters) [reply src/mutations/circle/putCircle.ts:59]: That should work as well. 👌🏻

File: src/mutations/circle/putCircle.ts [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+  switch (action) {
+    case ACTION.add: {
+      // checks: valid amount, duplicate circle
+      if (!trimedName || !trimedDisplayName) {
+        throw new UserInputError(
+          'circleName and displayName is required for creation'
+        )
+      }
+      if (amount < 0) {
```
[the remark this answers is not in the document]
Comment (devformatters) [reply src/mutations/circle/putCircle.ts:42]: 😉 Another checker! I also found other payment related resolver has this check too.

File: src/mutations/circle/toggleCircleSubscription.ts [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    case ACTION.subscribe: {
+      if (item) {
+        throw new DuplicateCircleSubscribeError('circle subscribed alraedy')
+      }
+
+      // re-init subscription if it doesn't exist
+      if (!subscription) {
+        const stripeSubscription = await paymentService.stripe.createSubscription(
+          {
+            customer: environment.stripeCustomerId,
```
[the remark this answers is not in the document]
Comment (devformatters) [reply src/mutations/circle/toggleCircleSubscription.ts:85]: Yes. It needs account setup.

File: src/connectors/stripe/index.ts [hunk 9]
```diff
@@ -138,6 +138,98 @@ class StripeService {
...
+    currency: PAYMENT_CURRENCY
+    interval: 'month'
+    productId: string
+  }) => {
+    try {
+      const price = await this.stripe.prices.create({
+        currency,
+        product: productId,
+        recurring: { interval },
+        unit_amount: amount,
```
Comment (robertu7) [inline src/connectors/stripe/index.ts:179]: `toProviderAmount(amount)`

**Review by robertu7 — CHANGES_REQUESTED** [review]

File: src/mutations/circle/putCircle.ts [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+      // update stripe product name
+      if (data.name) {
+        await paymentService.stripe.updateProduct({
+          id: updatedCircle.providerProductId,
+          name: data.name,
+        })
+      }
+
+      // delete unused assets
```
[the remark this answers is not in the document]
Comment (devformatters) [reply src/mutations/circle/putCircle.ts:205]: Nice reminder. Will do in the following PR.

File: src/mutations/circle/toggleCircleSubscription.ts [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+  switch (action) {
+    case ACTION.subscribe: {
+      if (item) {
+        throw new DuplicateCircleSubscribeError('circle subscribed alraedy')
+      }
+
+      // re-init subscription if it doesn't exist
+      if (!subscription) {
+        const stripeSubscription = await paymentService.stripe.createSubscription(
```
[the remark this answers is not in the document]
Comment (devformatters) [reply src/mutations/circle/toggleCircleSubscription.ts:83]: Added billing_cycle_anchor.

File: src/mutations/circle/toggleCircleSubscription.ts [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+      const stripeItem = await paymentService.stripe.createSubscriptionItem({
+        price: price.id,
+        subscription: subscription.id,
+      })
+
+      if (!stripeItem) {
+        throw new ServerError('cannot retrieve stripe subscription item')
+      }
+
+      await knex.transaction(async (trx) => {
```
[the remark this answers is not in the document]
Comment (devformatters) [reply src/mutations/circle/toggleCircleSubscription.ts:112]: Add a reaction to error catcher.

File: src/queries/circle/isMember.ts (hunk 3 above)
Comment (devformatters) [reply src/queries/circle/isMember.ts]: Fixed.

File: src/connectors/stripe/index.ts (hunk 9 above)
Comment (devformatters) [reply src/connectors/stripe/index.ts]: Added. Thanks.

File: src/queries/user/joinedCircles.ts [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+  { dataSources: { atomService }, knex }
+) => {
+  if (!id) {
+    return []
+  }
+
+  const circleIds = await knex
+    .select('price.circle_id')
+    .from('circle_subscription_item as csi')
+    .innerJoin('price', 'pirce.id', 'csi.price_id')
```
Comment (robertu7) [inline src/queries/user/joinedCircles.ts:16]: only items attached to the active subscription?

File: src/queries/circle/isMember.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+  _,
+  { viewer, dataSources: { atomService }, knex }
+) => {
+  if (!viewer.id) {
+    return false
+  }
+
+  const records = await knex
+    .select()
+    .from('circle_subscription_item as csi')
```
Comment (robertu7) [inline src/queries/circle/isMember.ts:15]: only items attached to the active subscription?

**Review by robertu7 — CHANGES_REQUESTED** [review]

File: src/common/utils/validator.ts [hunk 15]
```diff
@@ -93,3 +93,16 @@ export const isEnglish = (str: string): boolean => {
...
+
+/**
+ * Validate circle name. It only accepts alphabets, chinese characters and _.
+ */
+export const isValidCircleName = (name: string, size = 20): boolean => {
+  // TODO: add common disallowed names
+  if (!name || name.length < 2 || name.length > size) {
+    return false
+  }
+  return /^[_A-Za-z\u4E00-\u9FFF\u3400-\u4DFF\uF900-\uFAFF\u2e80-\u33ffh]*$/.test(
```
Comment (robertu7) [inline src/common/utils/validator.ts:105]: `/^[a-zA-Z0-9]*$/`?

**Review by robertu7 — CHANGES_REQUESTED** [review]

File: src/connectors/stripe/index.ts [hunk 16]
```diff
@@ -138,6 +138,108 @@ class StripeService {
...
+      const product = await this.stripe.products.update(id, { name })
+      return product
+    } catch (error) {
+      this.handleError(error)
+    }
+  }
+
+  deleteProduct = async ({ id }: { id: string }) => {
+    try {
+      await this.stripe.products.del(id)
```
[the remark this answers is not in the document]
Comment (devformatters) [reply src/connectors/stripe/index.ts:165]: But price has no method for deletion. 🤭

File: src/common/utils/validator.ts (hunk 15 above)
Comment (devformatters) [reply src/common/utils/validator.ts:105]: Number is not allowed written on Figma. 😂

File: src/mutations/circle/subscribeCircle.ts [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+    subscription = await atomService.create({
+      table: 'circle_subscription',
+      data: {
+        providerSubscriptionId: stripeSubscription.id,
+        userId: viewer.id,
+      },
+    })
+  }
+
+  // create stripe subscroption item
```
[the remark this answers is not in the document]
Comment (devformatters) [reply src/mutations/circle/subscribeCircle.ts]: Corrected.

File: src/queries/user/joinedCircles.ts (hunk 13 above)
Comment (devformatters) [reply src/queries/user/joinedCircles.ts]: Added.

File: src/queries/circle/isMember.ts (hunk 14 above)
Comment (devformatters) [reply src/queries/circle/isMember.ts:15]: Added.

Comment (devformatters) [reply src/queries/circle/isMember.ts:15]: Added.

File: src/common/utils/validator.ts (hunk 15 above)
Comment (robertu7) [reply src/common/utils/validator.ts:105]: @devformatters I saw displayName is reused by user and circle, and the validation rule of circle name is also similar to username?

**Review by robertu7 — CHANGES_REQUESTED** [review]

File: src/connectors/stripe/index.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (devformatters) [reply src/connectors/stripe/index.ts:165]: After discussion, let's put it aside for a while. 🦊

File: src/mutations/circle/subscribeCircle.ts [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+  const stripeItem = await paymentService.stripe.createSubscriptionItem({
```
[the remark this answers is not in the document]
Comment (devformatters) [reply src/mutations/circle/subscribeCircle.ts:96]: Fixed it.

**Review by robertu7 — APPROVED** [review]
Nice hard work! 

One last comment I left before, [please view](https://github.com/thematters/matters-server/pull/1673/files#r551705578).

**Pull request merged**